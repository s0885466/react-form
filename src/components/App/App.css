.text-area {
    transition: 2s;
}

.hide {
    display: none;
}

.show {
    display: block;
}