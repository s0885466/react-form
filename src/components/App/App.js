import React, {Component} from 'react';
import ProgressBar from '../Progressbar';
import InputArea from '../InputArea';
import Submit from '../Submit';
import checkModel from '../../Models/CheckModel';

import './App.css';

export default class App extends Component {
    state = {
        progressBar: {percent: 0},
        submit: {disabled: true},
        textArea: {class: 'hide'},
        inputs: [
            {
                label: 'Email', id: 'email', type: 'text',
                placeholder: '0885466@gmail.com', key: 1, check: false, value: ''
            },
            {
                label: 'Password', id: 'password', type: 'password',
                placeholder: '******', key: 2, check: false, value: ''
            },
            {
                label: 'Address', id: 'address', type: 'text',
                placeholder: 'Гаккелевская 27/2 кв.173', key: 3, check: false, value: ''
            },
            {
                label: 'Phone', id: 'phone', type: 'text',
                placeholder: '8 911 088 54 66', key: 4, check: false, value: ''
            }
        ]
    };

    checkInput = (id, value) => {

        this.setState(({inputs}) => {
            const index = inputs.findIndex(el => el.id === id);
            const newInputs = [...inputs];
            newInputs[index].value = value;
            newInputs[index].check = checkModel[id](value);
            const newPercent = this.changePercent(newInputs);
            const newDisabled = this.changeSubmitDisabled(newPercent);
            return {
                inputs: newInputs,
                submit: {disabled: newDisabled},
                progressBar: {percent: newPercent},
                textArea: {class: 'hide'}
            };
        });

    };

    onSubmitClick = () => {
        this.setState(({inputs}) => {
            const newInputs = [...inputs];
            newInputs.forEach(input => {
                input.value = '';
                input.check = false;
            });
            return {
                progressBar: {percent: 0},
                inputs: newInputs,
                textArea: {class: 'show'}
            }
        });
    };

    changePercent = (inputs) => {
        const len = inputs.length;
        const newPercent = Math.floor(inputs.filter(el => el.check === true).length * 100 / len);
        return newPercent;
    };

    changeSubmitDisabled = (percent) => {
        return percent !== 100;
    };

    render() {
        const {progressBar, inputs, submit, textArea} = this.state;

        return (
            <form>
                <ProgressBar percent={progressBar.percent} addPercent={this.addPercent}/>
                <InputArea inputs={inputs} checkInput={(id, value) => this.checkInput(id, value)}/>
                <Submit submit={submit} onSubmitClick={this.onSubmitClick}/>
                <div className={`text-center ${textArea.class} text-area` }>Форма отправлена, наши менеджеры связжутся с Вами</div>
            </form>
        );
    }
}
