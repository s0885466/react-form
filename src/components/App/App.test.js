import React from 'react';
import App from './App';
import {shallow} from 'enzyme';

const app = shallow(<App/>);

describe('Testing state <App/>', () => {

    it('<App/> have rendered correctly', () => {
        expect(app).toMatchSnapshot();
    });

    it('inputs state inputs is array', () => {
        expect(app.state().inputs).toBeArray();
    });

    it('submit state is object', () => {
        expect(app.state().submit).toBeObject();
    });

    it('textArea state is object', () => {
        expect(app.state().textArea).toBeObject();
    });

    it('progressBar state is object', () => {
        expect(app.state().progressBar).toBeObject();
    });
});

describe('testing onSubmitClick() <App/>', () => {
    it('testing onSubmitClick()', () => {
        app.instance().onSubmitClick();
        expect(app.state().progressBar).toEqual({percent: 0});
        expect(app.state().textArea).toEqual({class: 'show'});
    });
});

describe('testing checkInput(id, value)', () => {
    const check = (id, str, check) => {
        app.instance().checkInput(id, str);
        let index = app.state().inputs.findIndex(el => el.id === id);
        expect(app.state().inputs[index].check).toBe(check);
    };

    it('testing checkInput id=email', () => {
        check('email', 'dfsfsfsfs@f' , true);
        check('email', 'dfsfsfddsf' , false);
    });

    it('testing checkInput id=phone', () => {
        check('phone', '89110885466' , true);
        check('phone', '323424' , false);
    });

});