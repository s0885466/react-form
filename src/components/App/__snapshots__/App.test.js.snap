// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <App/> <App/> have rendered correctly 1`] = `ShallowWrapper {}`;

exports[`Testing state <App/> <App/> have rendered correctly 1`] = `ShallowWrapper {}`;
